File: static/solutions/java/src/main/java/com/epi/AddingCredits.java
package com.epi;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.NavigableMap;
import java.util.Set;
import java.util.TreeMap;

public class AddingCredits {
  // @include
  public static class ClientsCreditsInfo {
    private int offset = 0;
    private Map<String, Integer> clientToCredit = new HashMap<>();
    private NavigableMap<Integer, Set<String>> creditToClients
        = new TreeMap<>();

    public void insert(String clientID, int c) {
      remove(clientID);
      clientToCredit.put(clientID, c - offset);
      Set<String> set = creditToClients.get(c - offset);
      if (set == null) {
        set = new HashSet<>();
        creditToClients.put(c - offset, set);
      }
      set.add(clientID);
    }

    public boolean remove(String clientID) {
      Integer clientCredit = clientToCredit.get(clientID);
      if (clientCredit != null) {
        creditToClients.get(clientCredit).remove(clientID);
        if (creditToClients.get(clientCredit).isEmpty()) {
          creditToClients.remove(clientCredit);
        }
        clientToCredit.remove(clientID);
        return true;
      }
      return false;
    }

    public int lookup(String clientID) {
      Integer clientCredit = clientToCredit.get(clientID);
      return clientCredit == null ? -1 : clientCredit + offset;
    }

    public void addAll(int C) { offset += C; }

    public String max() {
      return creditToClients.isEmpty()
          ? ""
          : creditToClients.lastEntry().getValue().iterator().next();
    }
  }
  // @exclude

  public static void main(String[] args) {
    ClientsCreditsInfo a = new ClientsCreditsInfo();
    assert(a.max().isEmpty());
    String foo = "foo";
    String bar = "bar";
    String widget = "widget";
    String dothis = "dothis";
    String xyz = "xyz";
    String dd = "dd";
    assert(!a.remove(foo));
    a.insert(foo, 10);
    a.insert(foo, 1);
    a.insert(bar, 2);
    a.addAll(5);
    a.insert(widget, 3);
    a.addAll(5);
    a.insert(dothis, 4);
    assert(11 == a.lookup(foo));
    assert(12 == a.lookup(bar));
    assert(8 == a.lookup(widget));
    assert(4 == a.lookup(dothis));
    assert(a.remove(foo));
    assert(-1 == a.lookup(foo));
    assert(a.max().equals(bar));
    a.insert(xyz, 13);
    assert(a.max().equals(xyz));
    a.insert(dd, 15);
    assert(a.max().equals(dd));
  }
}

File: solutions/cpp/adding_credits.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <map>
#include <string>
#include <unordered_map>
#include <unordered_set>

using std::map;
using std::string;
using std::unordered_map;
using std::unordered_set;

// @include
class ClientsCreditsInfo {
 public:
  void Insert(const string& client_id, int c) {
    Remove(client_id);
    client_to_credit_.emplace(client_id, c - offset_);
    credit_to_clients_[c - offset_].emplace(client_id);
  }

  bool Remove(const string& client_id) {
    auto credit_iter = client_to_credit_.find(client_id);
    if (credit_iter != client_to_credit_.end()) {
      credit_to_clients_[credit_iter->second].erase(client_id);
      if (credit_to_clients_[credit_iter->second].empty()) {
        credit_to_clients_.erase(credit_iter->second);
      }
      client_to_credit_.erase(credit_iter);
      return true;
    }
    return false;
  }

  int Lookup(const string& client_id) const {
    auto credit_iter = client_to_credit_.find(client_id);
    return credit_iter == client_to_credit_.cend()
               ? -1
               : credit_iter->second + offset_;
  }

  void AddAll(int C) { offset_ += C; }

  string Max() const {
    auto iter = credit_to_clients_.crbegin();
    return iter == credit_to_clients_.crend() || iter->second.empty()
               ? ""
               : *iter->second.cbegin();
  }

 private:
  int offset_ = 0;
  unordered_map<string, int> client_to_credit_;
  map<int, unordered_set<string>> credit_to_clients_;
};
// @exclude

int main(int argc, char* argv[]) {
  ClientsCreditsInfo a;
  assert(a.Max() == "");
  assert(!a.Remove("foo"));
  a.Insert("foo", 10);
  a.Insert("foo", 1);
  a.Insert("bar", 2);
  a.AddAll(5);
  a.Insert("widget", 3);
  a.AddAll(5);
  a.Insert("dothis", 4);
  assert(11 == a.Lookup("foo"));
  assert(12 == a.Lookup("bar"));
  assert(8 == a.Lookup("widget"));
  assert(4 == a.Lookup("dothis"));
  assert(a.Remove("foo"));
  assert(-1 == a.Lookup("foo"));
  assert(a.Max().compare("bar") == 0);
  a.Insert("xyz", 13);
  assert(a.Max().compare("xyz") == 0);
  a.Insert("dd", 15);
  assert(a.Max().compare("dd") == 0);
  return 0;
}
